### ISSUE TITLE
StoreClient grpc invocation failure

### ISSUE THREAD

**Issue by dantengsky** [datafuselabs/datafuse#593]
**Describe the bug**
In each MySQL (**new**)client session, the first invocation of `create database xxx` works well, but subsequent  invocations of `create database xxx` lead to runtime exceptions like these:

~~~
ERROR fuse_query::servers::mysql::endpoints::endpoint_on_query] OnQuery Error: Code: 1002, displayText = status: Unknown error: Service was not ready: transport error: buffer's worker closed unexpectedly
~~~

I can confirm this is a client-side error.

The same problem exists while creating (remote) tables or appending data to storage (by using store client).

**How to reproduce**

~~~
>  make runquery
~~~

~~~
>  make runstore
~~~

~~~
> mysql -h127.0.0.1 -P3307  << EOF
create database t1;
create database t2;
EOF
~~~

Query log says:
~~~
INFO  common_flights::store_client] do_action: resp: {"CreateDatabase":{"database_id":6}}
ERROR fuse_query::servers::mysql::endpoints::endpoint_on_query] OnQuery Error: Code: 1002, displayText = status: Unknown error: Service was not ready: transport error: buffer's worker closed unexpectedly
~~~

The first `create database` statement is successfully executed, for each new MySQL client session.

**Which FuseQuery version to use**

  master branch (commit ea10471) 

**Additional context**

By a quick code browsing, I've noticed that StoreClient is lazily inited and cached by Datasource. Thus at least in the same session, the grpc client will be re-used.

I think that the error `Service was not ready: transport error` may be caused by a "mixture of runtime". (lazy init and clone of grpc client would be ok).

If I get it right, a grpc client is not guaranteed to be executed by the same tokio runtime in each session, which may be "the root cause".

https://github.com/datafuselabs/datafuse/blob/3e9da2ee47913511141130b9358645129af96640/fusequery/query/src/servers/mysql/mysql_handler.rs#L101-L113

**Comment by dantengsky** [comment]
I confirm that using the same tokio runtime for each query can fix this issue.

see commit d16503f

BUT, I am **not** suggesting that we should resolve this problem this way.

**Comment by BohuTANG** [comment]
Cool.
It looks make sense to use session runtime for the data pull.
But we need check ClickHouse client progress is working properly.

**Comment by dantengsky** [comment]
Failed to test by using a clickhouse client built from source:

### version of clickhouse-client
~~~
> ~/watch/ClickHouse/build/programs/clickhouse-client --version
ClickHouse client version 21.6.1.1.
~~~

### Not working (simple case)
Maybe the version does not match the crate `clickhouse-srv` well, or I did not config the server correctly, I failed to execute basic  SQL by using clickhouse client:

~~~
> clickhouse-client --host 127.0.0.1 --port 9000
ClickHouse client version 21.6.1.1.
Connecting to 127.0.0.1:9000 as user default.
Connected to datafuse server version 2021.5.0 revision 54405.

ClickHouse client version is older than ClickHouse server. It may lack support for new features.

datafuse :) select * from numbers(1);

SELECT *
FROM numbers(1)

Query id: 6ec41471-303c-4de7-adcf-6558e6f7bc4d


Exception on client:
Code: 261. DB::Exception: Unknown BlockInfo field number: 96: while receiving packet from 127.0.0.1:9000

Connecting to 127.0.0.1:9000 as user default.
Connected to datafuse server version 2021.5.0 revision 54405.

ClickHouse client version is older than ClickHouse server. It may lack support for new features.
~~~

### Not working (Create two databases in one session)
~~~


> ~/watch/ClickHouse/build/programs/clickhouse-client --host 127.0.0.1 --port 9000 -n << EOF
create database t1;
create database t2;
EOF
~~~

Logs of Quer Server

~~~
[2021-05-22T13:26:47Z INFO  common_flights::store_client] do_action: resp: {"CreateDatabase":{"database_id":18}}
[2021-05-22T13:26:52Z ERROR fuse_query::servers::clickhouse::clickhouse_handler] Error: Other("connection reset by peer")
[2021-05-22T13:26:52Z INFO  common_flights::store_client] do_action: resp: {"CreateDatabase":{"database_id":19}}
[2021-05-22T13:26:52Z ERROR fuse_query::servers::clickhouse::clickhouse_handler] Error: Other("connection reset by peer")
~~~

Seems that during executing the first statement,  clickhouse-client lost its connection to query-server, the second statement is executed by another session.

**Comment by BohuTANG** [comment]
ok.
 IMO should test ClickHouse with your patch~

**Comment by dantengsky** [comment]
Re-arrangement the runtimes of plan execution are risky or at least tricky, lots of things should be taken into account, I think.

I suggest that we temporarily fix this issue by creating a new StoreClient for each store operation, which I am quite sure can solve this problem (...Service was not ready: transport error..), and check this issue later, after we did integrate the basic functionalities of Store and Query.

**Comment by dantengsky** [comment]
> ok.
> IMO should test ClickHouse with your patch~

Commit d16503f only changes the internal behavior of  **mysql_handler.rs**,  "clickhouse_handler.rs" remains the same.

**Comment by BohuTANG** [comment]
Thanks @dantengsky 
All is ok to fix this issue in your any patch if you feel appropriate.

**Comment by zhang2014** [comment]
This problem also exists in FuseQuery. The reason is that the channel will call `tokio::spawn`(use the current runtime to schedule the query request. When the current runtime is destroyed, the channel will also be destroyed, so we need a long-time runtime. This may be global context, not session context)

reference: 
https://github.com/tower-rs/tower/blob/master/tower/src/buffer/service.rs#L76
https://github.com/datafuselabs/datafuse/pull/469/files#diff-1befb0f006ee42d05af965e9ef3f96c8073851f04b5709723108fee0c9249c71R159

**Comment by dantengsky** [comment]
> This problem also exists in FuseQuery. The reason is that the channel will call `tokio::spawn`(use the current runtime to schedule the query request. When the current runtime is destroyed, the channel will also be destroyed, so we need a long-time runtime. This may be global context, not session context)
> 
> reference:
> https://github.com/tower-rs/tower/blob/master/tower/src/buffer/service.rs#L76
> https://github.com/datafuselabs/datafuse/pull/469/files#diff-1befb0f006ee42d05af965e9ef3f96c8073851f04b5709723108fee0c9249c71R159

Thanks for the reference!

**Comment by dantengsky** [comment]
> Thanks @dantengsky
> All is ok to fix this issue in your any patch if you feel appropriate.

OK

**Comment by BohuTANG** [comment]
@zhang2014 @dantengsky 

After your information, seems that we'd have a global fight client pool.

**Comment by zhang2014** [comment]
> the disadvantage is:
maybe some tasks may often run first, others in low priority will be block until the firsts task finished.

The channel only accept requests and build a response future (in fact, the task is still running at query tokio runtime).

It is very flexible, which means that all requests will enter a runtime to queue (waiting for the runtime to approve the execution request). When we use session runtime to execute it, it means that all flight requests in this session need to be queued (but in fact, only one query is executing in one session at the same time).

reference:  https://github.com/tower-rs/tower/blob/master/tower/src/buffer/service.rs#L49

**Comment by zhang2014** [comment]
This may be very helpful for limited node load. But session runtime is OK for now(In fact, it may not different from use query runtime).

**Comment by BohuTANG** [comment]
> This may be very helpful for limited node load. But session runtime is OK for now(In fact, it may not different from use query runtime).

Thanks.
Is it a part of ISSUES-468 or another patch?

**Comment by drmingdrmer** [comment]
> I confirm that using the same tokio runtime for each query can fix this issue.
> 
> see commit [d16503f](https://github.com/datafuselabs/datafuse/commit/d16503f327ae2d1a6afc0291903962399ce7daba)
> 
> BUT, I am **not** suggesting that we should resolve this problem this way.

Hi guys, for now, are you gonna merge this commit d16503f to get one less thing to worry about? 🤔

**Comment by BohuTANG** [comment]
OK, if it blocks us let's resolve it.
@dantengsky open the PR please

**Comment by dantengsky** [comment]
This issue is fixed in PR #587, sorry for the late response
